kokkos: allow setting new gfx942_apu arch via apu variant

rbberger opened the pull request:
Add a ~unified_memory~ `apu` variant which promotes GPU architectures to their APU
equivalent. Right now this is just `gfx942` -> `gfx942_apu`.

masterleinad commented:
Why is this better than just specifying the unified memory architecture directly?

rbberger commented:
```
`gfx942_apu` is not a valid `amdgpu_target`
```

rbberger commented:
@spackbot fix style

masterleinad commented:
Would it make sense to make sure that specifying architectures other than gfx942 are mapped to something valid in Kokkos?

rbberger commented:
We could do that with a loop that adds conflicts for everything but what is in the supported map. Not sure what the status is of UM for Grace Hopper.

cedricchevalier19 commented:
I wanted to add that Grace-Hopper or even HMM can fall in this category.

I think a common variant can be convenient but:
- it is not what Kokkos does in its cmake options
- `unified_memory` can be confused with `uvm` or this kind of features

Do you know how others packages are dealing with AMD APU?

rbberger commented:
I think a lot of people are waiting on what KOKKOS will do. I only know of the unified_memory variant in hypre and umpire, but those might very well be UVM and have nothing to do with APUs.

The reason I spelled it out was to avoid the confusion between UM and UVM.

There is no clear direction from the AMD stack itself as far as I can tell.

rbberger commented:
Let me know if you want to rename the variant, e.g. to apu. The main purpose of this PR is so that people have something to test with during the next ElCap hackathon.

cedricchevalier19 reviewed (approved):
I think it is ok for now to enable tests on MI300A.
We can still think about renaming the option before the next spack release and we can extend it to Grace-Hopper or HMM in another PR.

rbberger commented:
Note, I haven't merged this one in yet since I didn't get the sense we've reached a consensus on how to proceed.

JBludau commented:
@rbberger would it help to meet to discuss this?

rbberger commented:
I'm currently at the El Cap hackathon at Sandia. AFAIK there's a conference call later this week?

cwpearson commented:
@rbberger could sit next to me to join the meeting - we could discuss it first and then let him go?

lenalopatina commented:
Our team at LANL is interested in this capability as well

rbberger committed:
kokkos: allow using new gfx942_apu arch

Add a apu variant which promotes GPU architectures to their APU
equivalent. Right now this is just gfx942 -> gfx942_apu.

rbberger commented:
@lucbv does this look good to you?

nmm0 reviewed (approved):
Thanks Richard!

lucbv reviewed (approved):
Looks good to me, thanks!
